### PR TITLE
[AArch64] Resolve FIXME: Use scavengeRegisterBackwards to find the best unused register

### PR DESCRIPTION
**Pull request by AZero13**

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64SpeculationHardening.cpp [hunk 1]
```diff
@@ -296,17 +296,16 @@ bool AArch64SpeculationHardening::instrumentControlFlow(
...
-    // FIXME: The below just finds *a* unused register. Maybe code could be
-    // optimized more if this looks for the register that isn't used for the
-    // longest time around this place, to enable more scheduling freedom. Not
-    // sure if that would actually result in a big performance difference
-    // though. Maybe RegisterScavenger::findSurvivorBackwards has some logic
-    // already to do this - but it's unclear if that could easily be used here.
-    Register TmpReg = RS.FindUnusedReg(&AArch64::GPR64commonRegClass);
+      TmpReg = RS.scavengeRegisterBackwards(AArch64::GPR64commonRegClass, I,
+                                            false, 0, false);
+    }
```
Comment (jayfoad) [inline llvm/lib/Target/AArch64/AArch64SpeculationHardening.cpp:308]: I think you can call scavengeRegisterBackwards even at the start of a block - can't you?
```
    if (I == MBB.begin())
      RS.enterBasicBlock(MBB);
    else
      RS.backward(I);
    TmpReg = RS.scavengeRegisterBackwards(AArch64::GPR64commonRegClass, I,
                                          false, 0, false);
```
The only reason this code was special casing `I == MBB.begin()` in the first place is for efficiency, not for correctness. The assumption is that `enterBasicBlock` is more efficient than `backward` which might have to step over arbitrarily many instructions.

File: llvm/test/CodeGen/AArch64/speculation-hardening.mir [hunk 2]
```diff
@@ -11,34 +12,148 @@
 # - other direct branches don't seem to be generated by the AArch64 codegen
 --- |
   define void @nobranch_fallthrough(i32 %a, i32 %b) speculative_load_hardening {
+  ; CHECK-LABEL: nobranch_fallthrough:
```
Comment (qcolombet) [inline llvm/test/CodeGen/AArch64/speculation-hardening.mir:15]: If you use the autogenerated test, make sure to clear the all check lines.
As is, there are labels that are repeated which may lead in weird failures with FileCheck. (e.g., the label nobranch_fallthrough appears twice.)

File: llvm/test/CodeGen/AArch64/speculation-hardening.mir [hunk 3]
```diff
@@ -11,34 +12,148 @@
 # - other direct branches don't seem to be generated by the AArch64 codegen
 --- |
   define void @nobranch_fallthrough(i32 %a, i32 %b) speculative_load_hardening {
```
Comment (qcolombet) [inline llvm/test/CodeGen/AArch64/speculation-hardening.mir:14]: Could you add a comment of what needs to happen or must not happen in the comment?

The manual checks were conveying this in a more direct way (but broken in its own way :P).

**Review by arsenm — CHANGES_REQUESTED** [review]
Previous comments

File: llvm/lib/Target/AArch64/AArch64SpeculationHardening.cpp [hunk 4]
```diff
@@ -296,17 +296,9 @@ bool AArch64SpeculationHardening::instrumentControlFlow(
...
-    if (I == MBB.begin())
-      RS.enterBasicBlock(MBB);
-    else
-      RS.backward(I);
-    // FIXME: The below just finds *a* unused register. Maybe code could be
-    // optimized more if this looks for the register that isn't used for the
-    // longest time around this place, to enable more scheduling freedom. Not
-    // sure if that would actually result in a big performance difference
-    // though. Maybe RegisterScavenger::findSurvivorBackwards has some logic
-    // already to do this - but it's unclear if that could easily be used here.
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/AArch64SpeculationHardening.cpp:308]: This comment should probably be reserved, fundamentally the scavenge is still doing the same thing and not finding a "nicer" register

**Pull request closed without merging**